### ISSUE TITLE
Interface implementation errors are not descriptive.

### ISSUE THREAD

**Issue by slimsag** [gopherjs/gopherjs#200]
Running [this program on play.golang.org](http://play.golang.org/p/dIjV7xedya) produces:

```
/tmp/sandbox161431897/main.go:14: cannot convert T literal (type T) to type FooBar:
	T does not implement FooBar (missing Bar method)
```

It tells us that `Bar` is the missing method (i.e. what the error actually is).

But running the same [program via gopherjs.org/play](http://www.gopherjs.org/play/#/nmf16uFkus) produces a error that is very hard to understand:

```
prog.go:14:13: cannot convert (T literal) (value of type T) to FooBar 
```

I actually have code that uses some very large composite interfaces -- and it's hard to track down exactly what I've done wrong (it's code that cannot be run via the standard Go compiler itself, only by gopherjs).

Any ideas on a potential workaround or improvement to this situation?

**Comment by shurcooL** [comment]
I haven't looked closely, but a _guess_ is that this may be a problem in one of internal Go packages, such as [`golang.org/x/tools/go/types`](https://godoc.org/golang.org/x/tools/go/types).

Given that the Go compiler is going to be written in Go soon, I'd expect there's not a large window for the Go packages to produce sub-par error messages compared to the previous C code. In any case, it might be worth looking into, and if that's the case, contributing to that package to make its error messages better would help both GopherJS and future Go compiler.

**Comment by slimsag** [comment]
If that's true, it'd be good to get a small reproducible program so I could report the issue at the appropriate place instead.

**Comment by shurcooL** [comment]
I think it is `go/types`. If you look at its [source](http://gotools.org/golang.org/x/tools/go/types), grep for `cannot convert %s to %s` and it comes up 3 times.

**Comment by shurcooL** [comment]
Hmm, I forgot that Go compiler is written in Go in 1.5, but it is currently still a largely automatically translated version from C.

It does not yet use `go/types` package. See https://github.com/golang/go/blob/25bf792197ff737fc84c8b1dd46be8a6b066333a/src/cmd/internal/gc/typecheck.go#L1759.

**Comment by neelance** [comment]
Yes, this error message comes from `go/types`, there is nothing I can do about it directly in GopherJS. The Go compiler is not yet using that package, but I guess at some point it will.

Here is a small test program to show the error:
```go
package main

import (
	"fmt"
	"go/ast"
	"go/parser"
	"go/token"

	"golang.org/x/tools/go/types"
)

func main() {
	fset := token.NewFileSet()

	f, err := parser.ParseFile(fset, "test.go", `
		package main

		type X interface {
			foo()
		}

		type S struct {}

		func main() {
			var x X
			x = &S{}
			_ = x
		}
	`, 0)
	if err != nil {
		panic(err)
	}

	_, err = types.Check("main", fset, []*ast.File{f})
	fmt.Println(err)
}
```

**Comment by neelance** [comment]
@slimsag Do you plan on reporting the issue to the Go issue tracker?

**Comment by slimsag** [comment]
@neelance apologies for the latent response. I've filled https://github.com/golang/go/issues/10260 just now.

**Comment by neelance** [comment]
Great, thanks!

**Comment by shurcooL** [comment]
The upstream issue golang/go#10260 has been resolved, but only in `go/types`, not `golang.org/x/tools/go/types`. So this issue in GopherJS will be resolved when #278 is merged and 1.6 comes out.

**Comment by shurcooL** [comment]
https://github.com/golang/go/issues/10260 has been resolved, we're now using `go/types` since #278 is merged, and Go 1.6 is out. So according to my comment above, this should be resolved.

However, I tested it, and it's not resolved. The error message is still:

```
$ go build
# github.com/shurcooL/play/31
./main.go:14: cannot convert T literal (type T) to type FooBar:
	T does not implement FooBar (missing Bar method)

$ gopherjs build
main.go:14:13: cannot convert (T literal) (value of type T) to FooBar
```

Where's the issue now?

**Comment by myitcv** [comment]
@shurcooL this is resolved for me locally:

```go
package main

type X interface {
	foo()
}

type S struct{}

func main() {
	var x X
	x = &S{}
	_ = x
}
```

```bash
$ gopherjs version
GopherJS 1.8-1
$ gopherjs build
main.go:11:6: cannot use &(S literal) (value of type *S) as X value in assignment: missing method foo
```

**Comment by shurcooL** [comment]
Thanks for posting an update here.

I get the same output for the program you posted above. However, I'm still seeing the same output as in https://github.com/gopherjs/gopherjs/issues/200#issuecomment-235015934 for the code in the original issue:

```Go
package main

type FooBar interface {
	Foo()
	Bar()
}

// T implements the FooBar interface -- but it's missing the Bar method (woops!).
type T struct{}

func (T) Foo() {}

func main() {
	_ = FooBar(T{})
}
```

```
$ go build
# github.com/shurcooL/play/31
./main.go:14: cannot convert T literal (type T) to type FooBar:
	T does not implement FooBar (missing Bar method)
$ gopherjs build
main.go:14:13: cannot convert (T literal) (value of type T) to FooBar
```

So the original issue doesn't seem to be resolved, right?

**Comment by myitcv** [comment]
@shurcooL you're absolutely right, sorry I didn't check that. 

I've just looked at the `go/types` source code. For assignments a reason is generated:

https://github.com/golang/go/blob/8db4d02e8fed0b42d71ef311c3b5481b2e72b26d/src/go/types/assignments.go#L60

but for conversions there is no reason:

https://github.com/golang/go/blob/8db4d02e8fed0b42d71ef311c3b5481b2e72b26d/src/go/types/conversions.go#L34

Looks like a relatively simple fix... will see if I can find the time over the weekend.

**Comment by nevkontakte** [comment]
This was fixed upstream in https://go-review.googlesource.com/c/go/+/359137.
